### PR TITLE
[Filebeat] add support for ndjson results to httpjson input

### PR DESCRIPTION
**Pull request by leehinman**
## What does this PR do?

adds support for ndjson results to httpjson input.

- if response Content-Type is "application/x-ndjson", collect the json
  objects in an array and assign array to key "ndjson-results" in a new
  json object
- if config response.force_ndjson is set, set response Content-Type
  header to "application/x-ndjson"
  
## Why is it important?

Some HTTP API endpoints return results as ndjson not json

## Checklist

- [x] My code follows the style guidelines of this project
~~- [ ] I have commented my code, particularly in hard-to-understand areas~~
- [x] I have made corresponding changes to the documentation
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

File: x-pack/filebeat/input/httpjson/internal/v2/pagination.go [hunk 1]
```diff
@@ -159,6 +160,13 @@ func (iter *pageIterator) getPage() (*response, error) {
 	r.url = *iter.resp.Request.URL
 	r.page = iter.n
 
+	if len(bodyBytes) > 0 && r.header.Get("Content-Type") == "application/x-ndjson" {
+		bodyBytes = append([]byte("{\"ndjson-results\":["), bodyBytes...)
+		bodyBytes = bytes.ReplaceAll(bodyBytes, []byte("\r\n"), []byte(","))
+		bodyBytes = bytes.ReplaceAll(bodyBytes, []byte("\n"), []byte(","))
+		bodyBytes = bytes.TrimRight(bodyBytes, ",")
+		bodyBytes = append(bodyBytes, []byte("]}")...)
+	}
```
Comment (marc-gr) [inline x-pack/filebeat/input/httpjson/internal/v2/pagination.go:169]:
```suggestion
	if len(bodyBytes) > 0 && r.header.Get("Content-Type") == "application/x-ndjson" {
		bodyBytes = append([]byte("["), bodyBytes...)
		bodyBytes = bytes.ReplaceAll(bodyBytes, []byte("\r\n"), []byte(","))
		bodyBytes = bytes.ReplaceAll(bodyBytes, []byte("\n"), []byte(","))
		bodyBytes = bytes.TrimRight(bodyBytes, ",")
		bodyBytes = append(bodyBytes, ']')
	}
```

Root arrays are supported and automatically split, so just converting the object stream to an array should make it work normally without needing to know about the special case if ndjson is received. WDYT?

File: x-pack/filebeat/docs/inputs/input-httpjson.asciidoc [hunk 2]
```diff
@@ -448,6 +448,27 @@ filebeat.inputs:
         value: '[[now (parseDuration "-1h")]]'
 ----
 
+[[response-force_ndjson]]
+[float]
+==== `response.force_ndjson`
+
+If the Content-Type header is set to "application/x-ndjson" then the ndjson results will be available in a new array called "ndjson-results".
```
Comment (marc-gr) [inline x-pack/filebeat/docs/inputs/input-httpjson.asciidoc:455]: I would add a comment about supporting ndjson at the top of the docs instead for visibility.

File: x-pack/filebeat/input/httpjson/internal/v2/response.go [hunk 3]
```diff
@@ -80,6 +83,10 @@ func (rp *responseProcessor) startProcessing(stdCtx context.Context, trCtx *tran
 	go func() {
 		defer close(ch)
 
+		if rp.forceNdjson {
+			resp.Header.Set("Content-Type", "application/x-ndjson")
```
Comment (marc-gr) [inline x-pack/filebeat/input/httpjson/internal/v2/response.go:87]: Is this just for testing purposes? If not, maybe there are other use cases for modifying response headers, and enabling writing to them in `transform_xxx.go` is a more general approach without adding specific config fields. WDYT?

**Review by marc-gr — COMMENTED** [review]
Really nice improvement!

Beside the couple of comments, would be nice to add some test for ndjson if possible.

File: x-pack/filebeat/input/httpjson/internal/v2/response.go (hunk 3 above)
Comment (leehinman) [reply x-pack/filebeat/input/httpjson/internal/v2/response.go:87]: Unfortunately not just for testing, some endpoints send back ndjson with content types of text/plain or application/json.  So we need some way to force ndjson code path.  That being said I like the idea of the general approach, let me give that a try.

Comment (andrewkroh) [reply x-pack/filebeat/input/httpjson/internal/v2/response.go]: After reading "if config response.force_ndjson is set, set response Content-Type
header to "application/x-ndjson" in the description, my first thought was that this sounded fairly specific. If we have several builtin handlers for different content types then maybe a generic overwrite header option would work and make future extension easier.

**Comment by marc-gr** [comment]
I opened https://github.com/elastic/beats/pull/23478, if it looks good to you, maybe we can merge it and then you just need to create a new ndjson decoder and use the `encode_as` option.

**Comment by leehinman** [comment]
yeah, I like the encode_as & decode_as stuff from #23478 better.  [Cunnighams Law at work](https://github.com/dwmkerr/hacker-laws#cunninghams-law). Closing this one.

**Pull request closed without merging**